### DISCUSSION TITLE
Getters and setters

### DISCUSSION THREAD

**Discussion by feniljariwala82** [adonisjs/core#2710] [Help] [answered]
I'm using Adonisjs v5. And I am trying to convert my JSON object to a "string" for storing in the SQLite database. I tried this, but setters and getters are not working. Anyone can help me out?

![json](https://user-images.githubusercontent.com/42708686/120266141-e961af00-c2be-11eb-9f06-3cab72f961ba.png)

**Comment by aayush123** [comment, marked as answer]
You can use model hooks to convert the data to & from JSON.
But I suggest a better option would be to use a computed column in your model.

**Reply by aayush123** [reply]
The following hook will convert object into JSON before saving a model:
```ts
  @beforeSave()
  public static async handleJSONBeforeSave(device: Device) {
    if (typeof device.detail === 'object') {
      device.detail = JSON.stringify(device.detail);
    }
  }
```

To retrieve the JSON as object for usage, you can:

1. Either add the following hooks to parse JSON
```typescript
  @afterSave()
  public static async handleJSONAfterSave(device: Device) {
    if (typeof device.detail === 'string') {
      device.detail = JSON.parse(device.detail);
    }
  }

  @afterFind()
  public static async handleJSONAfterFind(device: Device) {
    if (typeof device.detail === 'string') {
      device.detail = JSON.parse(device.detail);
    }
  }

  @afterFetch()
  public static async handleJSONAfterFetch(devices: Device[]) {
    devices.forEach((eachDevice) => {
      if (typeof eachDevice.detail === 'string') {
        eachDevice.detail = JSON.parse(eachDevice.detail);
      }
    });
  }
```

**OR**

2. Use the following computed column:

```ts
  @computed()
  public get parsedDetail() {
    const parsedDetail: any = JSON.parse(this.detail); // 👈 You can replace any with a type of your choice for nice autocomplete
    return parsedDetail;
  }
```

I personally prefer the computed column method. It's much cleaner and allows you to have a type for the parsed data.

**Reply by feniljariwala82** [reply]
Hooks worked for me, @aayush123 . I am really thankful for your advice and for sharing your knowledge with me. My code works fine now. Thank you

**Reply by aayush123** [reply]
No worries. Happy hacking! :)

**Comment by thetutlage** [comment]
Can you share how you set the `detail` property?

**Reply by feniljariwala82** [reply]
This is my **data** that I am trying to **seed** into the **database**.
 {
    device_id: cuid(),
    description: 'None',
    detail: { type: 'BR', make: '1.0', serial_no: 89798 },  **// json object**
    is_upload_enabled: false,
  },

**Detailed problem**: I want to **store** the 'detail' JSON object as a **string** in the SQLite database. And I want to **read** "detail" as a **JSON object** so that I can access keys and values in EDGE template.

-  And this is my seeder function

![seeder](https://user-images.githubusercontent.com/42708686/120272002-f506a300-c2c9-11eb-9aa3-b2f0e1480ca4.png)

**Error**:  Types of property 'detail' are incompatible.
    Type '{ type: string; make: string; serial_no: number; }' is not assignable to type 'string'.

- Model getter and setters
![column](https://user-images.githubusercontent.com/42708686/120272123-267f6e80-c2ca-11eb-953a-419755f2b41c.png)

**Comment by d4n1b** [comment]
The `column` decorator accepts a `consume` fn (reading from the DB) & `prepare` fn (writing to the DB)

![image](https://user-images.githubusercontent.com/17992718/120293721-71df5000-c2bd-11eb-9b27-65b85918f87d.png)

I'm not sure whether this is documented or not, but you can easily see it if you navigate through the types with your IDE.

I hope it helps
